spec.bs

```diff
@@ -2280,6 +2268,27 @@ To mend the intended distinction between [=top-level traversables=] and [=fenced
 Note: With these new definitions, a [=top-level traversable=] is essentially "unfenced" as described
 in the [[#nested-traversables-intro]].
 
+The algorithm as it exists today is a "fenced" algorithm, which will be re-introduced as a "fenced"
```

Comment:
Wait, why are we doing this anyways? None of this seems to be mentioned in the PR's OP.

Before this PR, existing algorithms in HTML that were non-prefixed, were "fenced" by default, and we introduced unfenced versions in our spec. But now I guess we're re-introducing (only one or two of them) as explicitly "fenced", and then I guess we also have our other "unfenced" version that our spec has introduced previously?

Mayybeee this is important, but I am not initially sold.

Reply:
The intention for this is to have an algorithm to get the "main frame" as well as get the "outermost main frame" (as it's called in the Chromium impl). The initial top-level traversable patch turns it into an "outermost main frame", but there's nothing left over to traverse up a frame tree until you reach the fenced frame root.

This is needed so that ad components sending automatic beacons can use the origin of the ad frame root. There are a couple alternatives to what I have here:

1. Turn this into a "get the ancestor fenced navigable" algorithm, and move it to the [nested traversables](https://wicg.github.io/fenced-frame/#nested-traversables-inner) section. Remove the `fenced top-level traversable` definition altogether.
2. Even though the implementation looks directly at the frame's origin rather than at the fenced frame config when determining the origin of the ad frame root, I think we could just put the ad root's origin in the fenced frame config for ad components. While a cross-origin navigation to the ad root would make that information out of sync, the cross-origin navigation will result in the new page losing access to ad components + the ad components being unloaded, so realistically there will be no behavioral differences between the spec and implementation.

Let me know which one you'd prefer.

Reply:
I'm thinking about this some more, and I don't think option 1 is feasible, since it'll become really awkward where it's being used. A couple more options:

3. Have "top-level traversable" take an optional "fenced" parameter that defaults to "false" (so this continues to be unfenced by default). But that still will be awkward to write out.
4. Write a "get the initiator ancestor root origin" algorithm and set `[=source snapshot params/initiator ancestor root origin=]` to the result of running that function. The algorithm will do the traversal manually. I think this will also be easier to understand, and we won't have to worry about making a distinction between fenced/unfenced top-level traversables; we can just keep all top-level traversables unfenced.

I'm going to go ahead and do option 4 since I think it'll be the cleanest way to handle this.